st2actions/st2actions/runners/paramiko_ssh_runner.py

```diff
@@ -0,0 +1,150 @@
+RUNNER_PRIVATE_KEY = 'private_key'
+RUNNER_PARALLEL = 'parallel'
+RUNNER_SUDO = 'sudo'
+RUNNER_ON_BEHALF_USER = 'user'
+RUNNER_REMOTE_DIR = 'dir'
+RUNNER_COMMAND = 'cmd'
+RUNNER_CWD = 'cwd'
+RUNNER_ENV = 'env'
+RUNNER_KWARG_OP = 'kwarg_op'
+RUNNER_TIMEOUT = 'timeout'
+RUNNER_SSH_PORT = 'port'
+
+
+class BaseParallelSSHRunner(ActionRunner, ShellRunnerMixin):
+
+    def __init__(self, runner_id):
+        super(BaseParallelSSHRunner, self).__init__(runner_id=runner_id)
+        self._hosts = None
+        self._parallel = True
+        self._sudo = False
+        self._on_behalf_user = None
+        self._username = None
+        self._password = None
+        self._private_key = None
+        self._kwarg_op = '--'
+        self._cwd = None
+        self._env = None
+        self._timeout = None
+        self._on_behalf_user = cfg.CONF.system_user.user
+
+        self._ssh_key_file = None
+        ssh_key_file = cfg.CONF.system_user.ssh_key_file
+        if ssh_key_file:
+            ssh_key_file = os.path.expanduser(ssh_key_file)
+            if os.path.exists(ssh_key_file):
+                self._ssh_key_file = ssh_key_file
+
+        self._parallel_ssh_client = None
+
+    def pre_run(self):
+        LOG.debug('Entering BaseParallelSSHRunner.pre_run() for liveaction_id="%s"',
+                  self.liveaction_id)
+        hosts = self.runner_parameters.get(RUNNER_HOSTS, '').split(',')
+        self._hosts = [h.strip() for h in hosts if len(h) > 0]
+        if len(self._hosts) < 1:
+            raise ActionRunnerPreRunError('No hosts specified to run action for action %s.',
+                                          self.liveaction_id)
+        self._username = self.runner_parameters.get(RUNNER_USERNAME, cfg.CONF.system_user.user)
+        self._username = self._username or cfg.CONF.system_user.user
+        self._password = self.runner_parameters.get(RUNNER_PASSWORD, None)
+        self._ssh_port = self.runner_parameters.get(RUNNER_SSH_PORT, 22)
+        self._private_key = self.runner_parameters.get(RUNNER_PRIVATE_KEY, self._ssh_key_file)
+        self._parallel = self.runner_parameters.get(RUNNER_PARALLEL, True)
+        self._sudo = self.runner_parameters.get(RUNNER_SUDO, False)
+        self._sudo = self._sudo if self._sudo else False
+        self._on_behalf_user = self.context.get(RUNNER_ON_BEHALF_USER, self._on_behalf_user)
+        self._cwd = self.runner_parameters.get(RUNNER_CWD, None)
+        self._env = self.runner_parameters.get(RUNNER_ENV, {})
+        self._kwarg_op = self.runner_parameters.get(RUNNER_KWARG_OP, '--')
+        self._timeout = self.runner_parameters.get(RUNNER_TIMEOUT,
+                                                   FABRIC_RUNNER_DEFAULT_ACTION_TIMEOUT)
+
+        LOG.info('[BaseParallelSSHRunner="%s", liveaction_id="%s"] Finished pre_run.',
+                 self.runner_id, self.liveaction_id)
+
+        concurrency = int(len(self._hosts) / 3) + 1 if self._parallel else 1
+
+        if self._password:
```

Comment:
Do we want to support password protected keys?

If we do, we need to allow user to specify both - password and key file (in case both are specified, password is used to unlock the key file).